File: 백준/Silver/14888. 연산자 끼워넣기/연산자 끼워넣기.java

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class Main {
    private static int size = 0;
    private static int MIN_RESULT = Integer.MAX_VALUE;
    private static int MAX_RESULT = Integer.MIN_VALUE;
    private static int[] numbers;
    private static final int[] operations = new int[4];

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        size = Integer.parseInt(br.readLine());
        numbers = new int[size];

        StringTokenizer st1 = new StringTokenizer(br.readLine());
        for (int i = 0; i < size; i++) {
            numbers[i] = Integer.parseInt(st1.nextToken());
        }

        StringTokenizer st2 = new StringTokenizer(br.readLine());
        for (int i = 0; i < 4; i++) {
            operations[i] = Integer.parseInt(st2.nextToken());
        }

        dfs(numbers[0], 1);

        System.out.println(MAX_RESULT);
        System.out.println(MIN_RESULT);
    }

    private static void dfs(int number, int depth) {
        if (depth == size) {
            MAX_RESULT = Integer.max(number, MAX_RESULT);
            MIN_RESULT = Integer.min(number, MIN_RESULT);
            return;
        }

        for (int i = 0; i < 4; i++) {
            if (operations[i] > 0) {
                operations[i]--;

                switch (i) {
                    case 0:
                        dfs(number + numbers[depth], depth + 1);
                        break;
                    case 1:
                        dfs(number - numbers[depth], depth + 1);
                        break;
                    case 2:
                        dfs(number * numbers[depth], depth + 1);
                        break;
                    case 3:
                        dfs(number / numbers[depth], depth + 1);
                        break;
                }

                operations[i]++;
            }
        }
    }
}

File: 백준/Silver/14888. 연산자 끼워넣기/연산자 끼워넣기.kt
package com.sb.framework

import java.io.BufferedReader
import java.io.InputStreamReader
import java.util.*
import kotlin.math.max
import kotlin.math.min

var count: Int = 0
var minResult = Int.MAX_VALUE
var maxResult = Int.MIN_VALUE
var operator: IntArray = IntArray(4)
lateinit var numbers: IntArray

fun main() {
    val br = BufferedReader(InputStreamReader(System.`in`))

    count = br.readLine().toInt()
    val inputNumbers = StringTokenizer(br.readLine())
    numbers = IntArray(count)

    for (i in 0 until count) {
        numbers[i] = inputNumbers.nextToken().toInt()
    }

    val numbersTokens = StringTokenizer(br.readLine())
    for (i in 0 until 4) {
        operator[i] = numbersTokens.nextToken().toInt()
    }

    dfs(numbers[0], 1)

    println(maxResult)
    println(minResult)
}

fun dfs(number: Int, nextIndex: Int) {
    if (nextIndex == count) {
        minResult = min(minResult, number)
        maxResult = max(maxResult, number)

        return
    }

    for (i in 0 until 4) {
        if (operator[i] > 0) {
            operator[i]--

            when (i) {
                0 -> dfs(number + numbers[nextIndex], nextIndex + 1)
                1 -> dfs(number - numbers[nextIndex], nextIndex + 1)
                2 -> dfs(number * numbers[nextIndex], nextIndex + 1)
                3 -> dfs(number / numbers[nextIndex], nextIndex + 1)
            }

            operator[i]++
        }
    }
}